rosdep/python.yaml

```
@@ -4595,6 +4595,9 @@ python-sympy:
     wily_python3: [python3-sympy]
     xenial: [python-sympy]
     xenial_python3: [python3-sympy]
+python-systemd:
+  debian: [python-systemd]
+  ubuntu: [python-systemd]
```

Since python-systemd is not available on forthcoming Debian and the latest Ubuntu platfom we can't elide the version here:

```suggestion
python-systemd:
  debian:
    stretch: [python-systemd]
    buster: [python-systemd]
    '*': null
  ubuntu:
    bionic: [python-systemd]
    xenial: [python-systemd]
    '*': null
```

---

That would be wrong.
Imagine a package.xml with the python-systemd dependency and you want to use it with Python 2:

`<exec_depend condition="$ROS_PYTHON_VERSION == 2">python-systemd</exec_depend>`

A rosdep rule `debian: [python-systemd]` states that the python-systemd package needs to be installed (you could get it from Debian buster or build it yourself for example), whereas `'*': null` states that the dependency is not needed which is untrue.

---

I have to go dig up the documentation but I had thought that a null definition essentially undefines a key whereas a definition with no installations would be an empty []. In any case there is a way to undefine a key so that it is an error on platforms where the key isn't available I just have to go find it.

---

but why do you think you can define that a key is not available (you can have it through other channels)?

---

> but why do you think you can define that a key is not available (you can have it through other channels)?

If I understand this question correctly, you're asking why it's okay to undefine this key when the package could be provided from a non-standard repository. If I'm mistaken in that interpretation please let me know.

In order to be accepted definitions must come from one of the sources described here: https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#expected-rosdep-sources. On Ubuntu those are the main, universe, and multiverse repositories and I've just seen that the Debian section is copy-pasted and invalid, so I'll open a PR to fix that.

If an alternative source is used to provide a package for a platform that doesn't have it in the expected sources, a custom set of rosdep definitions can be used to override and define that key for the platform.

---

So in this case, I believe the `null` is implied by the absence of a mapping. I'm moderately confident that any occurrence of `'*': null` is actually implied and explicitly adding it is superfluous.

With that said, if the package isn't available for all supported platforms, explicitly listing the supported platforms as @nuclearsandwich suggests results in a better error message than listing a package as available for a platform when it actually isn't. Instead of rosdep trying to install a package that doesn't exist and getting an error out of apt, rosdep will report that there isn't a package available to satisfy the key, which could save someone from hunting around in vain for which repository they think that they need to enable to get the package to install.

---

> I've just seen that the Debian section is copy-pasted and invalid

https://github.com/ros/rosdistro/pull/27236

---

>  I'm moderately confident that any occurrence of `'*': null` is actually implied and explicitly adding it is superfluous.

You're right. For some reason the vagaries of rosdep don't stick with me the way most computer trivia does. I've updated my suggestion to remove those.

---

Some other thoughts:

> null definition essentially undefines a key whereas a definition with no installations would be an empty []

I believe this is an accurate description of the difference between the behavior of `null` and `[]`, and I'll add that the absence of a valid mapping will produce the same behavior as `null` (not `[]`).

> ...states that the python-systemd package needs to be installed (you could get it from Debian buster or build it yourself for example)

In reference to "build it yourself" - if you have an alternate source of packages enabled on your system that you want rosdep to be able to resolve against, you can absolutely add your own keys and rules on top of the keys and rules in `rosdistro` by listing those sources under `/etc/ros/rosdep/sources.list.d`. That said, an unsatisfied key as reported by rosdep doesn't prevent you from building a package if the requirement is satisfied by some other means, so just invoking rosdep with `--skip-keys foo` to ignore a key that you know you have satisfied manually is a lot less work.